### PR TITLE
FOGL-8706 Implement a worker thread pool in the storage layer

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

File: docs/storage.rst [hunk 1]
```diff
@@ -69,7 +69,9 @@ user interface to set the storage engine and its options.
 
   - The *Readings Plugin* may be set to any of the above and may also be set to use the SQLite In Memory plugin by entering the value *sqlitememory* into the configuration field.
 
-  - The *Database threads* field allows for the number of threads used for database housekeeping to be controlled. In normal circumstances 1 is sufficient. If performance issues are seen this can be increased however it is rarely required to be greater than 1 and can have counter productive effects on heavily loaded systems.
+  - The *Storage API threads* field allows for the number of threads used by the Storage service API for handling incoming requests.
+
+  - The *Worker thread pool* is used for executing operations against the readings data with Fledge. The number of threads in this thread pool is control via this setting. Increasing the thread pool size will increase the number of operations the can be executed in parallel, however care should be exercised as increasing this will increase any contention issues within the storage layer and will become counter productive once the storage plugin is overloaded.
```
Comment (Mohit04tomar) [inline docs/storage.rst:74]: ...The number of threads in this pool is controlled via this setting...

File: C/services/storage/storage_api.cpp [hunk 2]
```diff
@@ -427,7 +445,8 @@ StorageApi *StorageApi::getInstance()
 {
 	if (m_instance == NULL)
 	{
-		m_instance = new StorageApi(0, 1);
+		Logger::getLogger()->warn("Creating a default storage API instance, tunign parameters willbe ignored");
```
Comment (Mohit04tomar) [inline C/services/storage/storage_api.cpp:448]: Creating a default Storage API instance, tuning parameters will be ignored

File: C/services/storage/storage_api.cpp [hunk 3]
```diff
@@ -548,7 +593,69 @@ void StorageApi::respond(shared_ptr<HttpServer::Response> response, const string
...
+			}
+			delete op;
+			lck.lock();
+		}
+		m_queueCV.wait(lck);
+	}
+}
 
+/**
+ * Append a request to the readingfs request queue
```
Comment (Mohit04tomar) [inline C/services/storage/storage_api.cpp:635]: Append a request to the readings request queue

File: C/services/storage/storage_api.cpp [hunk 4]
```diff
@@ -548,7 +593,69 @@ void StorageApi::respond(shared_ptr<HttpServer::Response> response, const string
...
+ *
+ * If the queu is startign to get long delay the return as
```
Comment (Mohit04tomar) [inline C/services/storage/storage_api.cpp:637]: If the queue is starting to get long delay the return as

**Review by Mohit04tomar — DISMISSED** [review]
@MarkRiddoch I have tested and reviewed the pull request, and no issue has been found, except some typos

**Review by pintomax — APPROVED** [review]

**Comment by Mohit04tomar** [comment]
@MarkRiddoch I think title of the pull request should be `FOGL-8706 Implement a worker thread pool in the storage layer `

**Pull request merged**